### PR TITLE
padded string memory leak

### PR DESCRIPTION
**Pull request by DBJDBJ**
Also -- get_corpus  standard implementation  +  moved to header

jsonioutil.cpp to be removed

### CODE REVIEW

**Comment by lemire** [comment]
Can you merge with master and report performance results with the new get_corpus_benchmark? Ideally, we would like better performance.


https://github.com/lemire/simdjson/pull/456

**Comment by DBJDBJ** [comment]
> Can you merge with master and report performance results with the new get_corpus_benchmark? Ideally, we would like better performance.
> 
> #456

1.  padded_string is actually streamlined (see the title?)  not `get_corpus()`. `get_corpus()` is just a side effect.
2. detrimental influence of exceptions on performance and size is [proven](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1640r1.html),  thus I conclude `simdjson` must not use exceptions. That also includes current `get_corpus()`.  this new 'get_corpus` uses much less exceptions **and** is very malleable to optimizations.
3. I have done as much as I can in my spare time.

**Comment by lemire** [comment]
Ok. I ran the benchmarks...

Current speeds...(master)
```
$ ./get_corpus_benchmark_master jsonexamples/gsoc-2018.json 
file size: 0.00309928 GB
average speed: 4.25454 GB/s
min speed    : 2.79942 GB/s
max speed    : 5.86556 GB/s
```


Speed after merging this PR:

```
$ ./get_corpus_benchmark jsonexamples/gsoc-2018.json 
file size: 0.00309928 GB
average speed: 0.364734 GB/s
min speed    : 0.332593 GB/s
max speed    : 0.398785 GB/s
```

So, as you can see, it comes with more than a 10x performance penalty.

**Comment by lemire** [comment]
I am pretty sure that the following is to be avoided at all cost if we care about performance...
```
   ifstream ifs_(filename.data());
   string str_(istreambuf_iterator<char>{ifs_}, {});
```

**Comment by DBJDBJ** [comment]
image

@lemire

I expected that. My bench is [here](https://github.com/dbj-systems/simdjson_research) . 
I have added (in there) "no exceptions" `get_corpus()` version.  I do think the immediate roadmap of `simdjson` is very simple 3 steps:

1.  do not use try/throw/catch
2. do not use <iostream>
3. do not output to terminal/console 

> So, as you can see, it comes with more than a 10x performance penalty.

This always happens when one is using std lib. Good, we have this benchmark. 

I emailed you a few days (weeks) ago suggesting re-developing in C11, and if not then being economical on std lib usage. I repeat: I favor C++ core language, not std lib. 

The focus of my last PR  is padded_string, not `get_corpus()`, I will leave `get_corpus()` as it was (i.e. unchanged as it was before) of course.

**Comment by DBJDBJ** [comment]
```cpp
  // move assignment
  padded_string &operator=(padded_string && other_ ) {
    using namespace std;
    std::swap(data_ptr, other_.data_ptr);
    std::swap(viable_size, other_.viable_size);
    // free the payload
    // remove this and you will have a mem leak
    simdjson::aligned_free_char(other_.data_ptr);
    other_.data_ptr = nullptr;
    other_.viable_size = 0;
    return *this;
  }
```
Usually closing the major mem leak makes "other things" pop-up elsewhere. On my WIN10 pro machine, all is fine beside two tests in the Release build. 
image

File: include/simdjson/padded_string.h [hunk 1]
```diff
@@ -24,7 +35,9 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
     return nullptr;
   }
 #endif // NDEBUG
-  memset(padded_buffer + length, 0, totalpaddedlength - length);
+  
+    // ? -> memset(padded_buffer + length, 0, totalpaddedlength - length);
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Why would you remove the memset?

Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: if not strictly required that will improve the performance. What do you think? Although on -O3 that `memset` should be instant. Perhaps.  I am not sure. Testing as we speak.

**Comment by DBJDBJ** [comment]
@lemire  this is all being done exactly like that, I have removed the "double free" etc .... give me a few more hours, I am testing it in isolation here ...  

In the mean time can you help me understand these two "failures"  ?

File: include/simdjson/jsonioutil.h [hunk 2]
```diff
@@ -17,16 +17,18 @@ namespace simdjson {
...
-// caller is responsible to free (aligned_free((void*)result.data())))
 //
 // throws an exception if the file cannot be opened, use try/catch
+//
+//      padded_string p{} ;
 //      try {
 //        p = get_corpus(filename);
 //      } catch (const std::exception& e) {
-//        aligned_free((void*)p.data());
+//          p.~padded_string() ;
```
Comment (lemire) [inline include/simdjson/jsonioutil.h]: Calling the destructor in such a way almost ensures that the destructor will be called twice.

File: include/simdjson/padded_string.h (hunk 1 above)
Comment (lemire) [reply include/simdjson/padded_string.h]: Have you determined that it is not required?

File: src/jsonioutil.cpp [hunk 3]
```diff
@@ -1,36 +1,36 @@
 #include "simdjson/jsonioutil.h"
-#include <cstdlib>
-#include <cstring>
-#include <climits>
+ #include <cstdlib>
```
Comment (lemire) [inline src/jsonioutil.cpp]: Please don't introduce unnecessary white-space changes. You have taken this one file and added a leading space to all but the first line. Please undo this change.

File: src/jsonstream.cpp [hunk 4]
```diff
@@ -3,6 +3,7 @@
 #include "simdjson/isadetection.h"
 #include "jsoncharutils.h"
 
+ #include <stdexcept>
```
Comment (lemire) [inline src/jsonstream.cpp]: Please put the standard headers first.

File: include/simdjson/padded_string.h [hunk 5]
```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
-  char *data() const  { return data_ptr; }
+  char *data() const { return data_ptr; }
 
 private:
+    // no copying
   padded_string &operator=(const padded_string &o) = delete;
   padded_string(const padded_string &o) = delete;
 
-  size_t viable_size;
+  size_t viable_size{};
```
Comment (lemire) [inline include/simdjson/padded_string.h]: This looks like an unnecessary change.

File: include/simdjson/padded_string.h [hunk 6]
```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
   }
 
-  size_t size() const  { return viable_size; }
+  size_t size() const { return viable_size; }
 
-  size_t length() const  { return viable_size; }
+  size_t length() const { return viable_size; }
 
-  char *data() const  { return data_ptr; }
+  char *data() const { return data_ptr; }
```
Comment (lemire) [inline include/simdjson/padded_string.h]: This looks like a white-space only change, please avoid those.

File: include/simdjson/padded_string.h [hunk 7]
```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
+      data_ptr = nullptr;
+      viable_size = size_t(0);
+    }
   }
 
-  size_t size() const  { return viable_size; }
+  size_t size() const { return viable_size; }
 
-  size_t length() const  { return viable_size; }
+  size_t length() const { return viable_size; }
```
Comment (lemire) [inline include/simdjson/padded_string.h]: This looks like a white-space only change, please avoid those.

File: include/simdjson/padded_string.h [hunk 8]
```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
   ~padded_string() {
+    if (data_ptr != nullptr ) {
       aligned_free_char(data_ptr);
+      data_ptr = nullptr;
+      viable_size = size_t(0);
+    }
   }
 
-  size_t size() const  { return viable_size; }
+  size_t size() const { return viable_size; }
```
Comment (lemire) [inline include/simdjson/padded_string.h]: This looks like a white-space only change, please avoid those.

File: include/simdjson/padded_string.h [hunk 9]
```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
+  padded_string b("B");
+       a.swap(b) ;
+         printf("%s", a.data() ); // "B"
+  */
+  void swap( padded_string &right_) {
+    // auto &left_ = *this;
+    std::swap(*this, right_);
   }
 
+  /*
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Please don't include commented dead code in a PR to be reviewed.

File: include/simdjson/padded_string.h [hunk 10]
```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
+    if (!ps_.reset(length_))
+      return ENOMEM;
+
+    std::rewind(fp);
+    size_t readb = std::fread(ps_.data(), 1, length_, fp);
+    std::fclose(fp);
+
+    if (readb != length_) {
+      // free mem before bad return
+      ps_.~padded_string();
```
Comment (lemire) [inline include/simdjson/padded_string.h]: I am really uneasy about explicit calls to the destructor. I don't think it is idiomatic.

File: include/simdjson/padded_string.h (hunk 5 above)
Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: we have been through this before :)  C++17 standard says if value init aka '{}' is not used variable/constant is initialized to indeterminate value and if opposite it is "zeroed" to "zero" as appropriate to the type in use.  And that is indeed the case for CL (aka MSVC).  

```cpp
// var is "zeroed" to "zero" for the type T
T var{} ;
```

File: include/simdjson/padded_string.h (hunk 10 above)
Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: removed already

File: src/jsonioutil.cpp (hunk 3 above)
Comment (DBJDBJ) [reply src/jsonioutil.cpp]: this file is to be removed

File: include/simdjson/jsonioutil.h (hunk 2 above)
Comment (DBJDBJ) [reply include/simdjson/jsonioutil.h]: already removed

File: include/simdjson/padded_string.h (hunk 1 above)
Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: Not sure ... here hitting head one differences between runtimes and compilers.  Leaving it in.

File: include/simdjson/padded_string.h (hunk 5 above)
Comment (lemire) [reply include/simdjson/padded_string.h]: I know the C++ syntax, but isn't viable_size initialize in the constructor?

I should be able to look at this line and understand why it was changed. As it is, I have no idea why you changed it.

By itself, I do not find it objectionable, but you are making many, many changes (hundreds of lines) and I am struggling to find the motivation for most of these changes. This makes it hard to review.

Please change as few lines as possible. 

Can you reorg. your PRs so that they are narrow and focused, with clearly stated benefits?

File: include/simdjson/padded_string.h [hunk 11]
```diff
@@ -17,6 +17,10 @@
 typedef int errno_t ;
 #endif // _MSC_VER
 
+#ifndef SIMDJSON_PADDING
```
Comment (lemire) [inline include/simdjson/padded_string.h]: SIMDJSON_PADDING is defined in common_defs.h. Instead of redefinining it again, why not just do

#include "simdjson/common_defs.h"

?

File: include/simdjson/padded_string.h [hunk 12]
```diff
@@ -46,34 +50,37 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
-  bool reset(size_t new_size_) noexcept 
+  static bool reset(type & pstring_, size_t new_size_) noexcept 
   {
-    this->~padded_string(); 
-    viable_size = new_size_;
-    data_ptr = allocate_padded_buffer(new_size_);
-    if (data_ptr != nullptr) {
-      data_ptr[new_size_] = '\0'; // easier when you need a c_str
+      if (pstring_.data())
+        pstring_.~type();
```
Comment (lemire) [inline include/simdjson/padded_string.h]: I don't think calling the destructor on a live object is idiomatic.

File: src/jsonstream.cpp [hunk 13]
```diff
@@ -1,9 +1,11 @@
 #include <map>
+#include <stdexcept>
+
 #include "simdjson/jsonstream.h"
 #include "simdjson/isadetection.h"
 #include "jsoncharutils.h"
 
-
+
```
Comment (lemire) [inline src/jsonstream.cpp]: Please avoid unnecessary white-space changes.

File: include/simdjson/padded_string.h [hunk 14]
```diff
@@ -33,77 +50,160 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
+       a.swap(b) ;
+         printf("%s", a.data() ); // "B"
+  */
+  void swap( padded_string &right_) {
+    std::swap(*this, right_);
   }
 
+  /*
+     padded_string a("A");
+     a.~padded_string() ;
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Calling the destructor directly is almost surely not idiomatic.

File: include/simdjson/padded_string.h [hunk 15]
```diff
@@ -33,77 +50,160 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
...
+  explicit padded_string(const char *data, size_t length) noexcept
+      : padded_string(length) 
+  {
     if ((data != nullptr) and (data_ptr != nullptr)) {
       memcpy(data_ptr, data, length);
-      data_ptr[length] = '\0'; // easier when you need a c_str
+      // probably not necessary
+      data_ptr[length] = '\0'; 
+    } else {
+      /* do what? */
```
Comment (lemire) [inline include/simdjson/padded_string.h]: The C++ thing to do would be to throw an exception. In our case, I recommend simply assuming that data_ptr might be nullptr and unallocated. I suggest setting the viable_length to zero.

Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: OK

File: include/simdjson/padded_string.h (hunk 14 above)
Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: hopefully, no users will do it ...

File: src/jsonstream.cpp (hunk 13 above)
Comment (DBJDBJ) [reply src/jsonstream.cpp]: OK

File: include/simdjson/padded_string.h (hunk 12 above)
Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: AFAK I am not doing it in this case. Advice: std::lib code is full of destructor calls.

File: include/simdjson/padded_string.h (hunk 11 above)
Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: ok will remove in next PR

File: include/simdjson/padded_string.h (hunk 5 above)
Comment (DBJDBJ) [reply include/simdjson/padded_string.h]: I am aware of that and will change that behaviour, so we do not indulge in a ping-pong around each PR :) I might do the issue first, discuss it and then do the PR on the agreed outcome of the issue discussion.

File: src/jsonstream.cpp (hunk 4 above)
Comment (DBJDBJ) [reply src/jsonstream.cpp]: done

**Comment by DBJDBJ** [comment]
@lemire major mem leak is found and closed ... can we merge this one now?

**Comment by lemire** [comment]
I am not aware of a memory leak in simdjson. Can you elaborate? Is this in relation with an issue what has been submited?

**Comment by lemire** [comment]
As it is, I have concerns with this PR. There are too many changes and it is not clear where the fix is and what the problem was.

**Comment by DBJDBJ** [comment]
> I am not aware of a memory leak in simdjson. Can you elaborate? Is this in relation to an issue that has been submitted?

yesterday I posted (also with a big image in the comment) ... have you not seen that?

image

Here is the fix once more:

```cpp
// padded_string  
// move assignment
  padded_string &operator=(padded_string && other_ ) {
    using namespace std;
    std::swap(data_ptr, other_.data_ptr);
    std::swap(viable_size, other_.viable_size);
    // free the payload
    // remove this and you will have a mem leak
    simdjson::aligned_free_char(other_.data_ptr);
    other_.data_ptr = nullptr;
    other_.viable_size = 0;
    return *this;
  }
```
So, for example ... Each time  `get_corpus()` was returning a padded_stream  instance that instance data was not freed ...

I assume you can create a branch to check for this.

**Comment by lemire** [comment]
Thank you.

**Comment by DBJDBJ** [comment]
> Thank you.

Happy to help. 

After this PR workflow will be: 1. create the issue, 2: implement what was decided in the issue chat.

**Comment by lemire** [comment]
Please see https://github.com/lemire/simdjson/pull/461

**Comment by DBJDBJ** [comment]
Summary of changes 

1. benchmark/get_corpus_benchmark.cpp -- not critical

2. extra/dumpbits.h  -- not critical (it seems that is not used anywhere?)

3. include/simdjson/jsonioutil.h  -- just moved the inline padded_string get_corpus(std::string_view filename) in here, so that jsonioutil.cpp can be removed

4. include/simdjson/padded_string.h -- no mem leak + optimized implementation  
    1. static errno_t load(type  &ps_, FILE *fp) noexcept ; is not used and it thus removed by the compiler, it looks useful, at lease to me

5. src/jsonioutil.cpp -- can be removed

6. src/jsonstream.cpp --  #include <stdexcept> was required

Let me know about fixes required for the failing CI checks.

**Comment by lemire** [comment]
I am going to close this. We have a memory leak and it needs to be fixed. There is another PR for this.

The rest of the changes appear to refer to style and/or refactoring. 

You refer to your other changes to padded_string as "optimizations". If this means "performance enhancement", then please provide benchmark results. If the "optimization" is about "code style that you prefer", then I am not going to merge it.

We are way behind with respect to our move toward version 0.3 and contributions that do not fix bugs or do not get us closer to the next release are discouraged/blocked.

**Comment by DBJDBJ** [comment]
You will get the required padded_string source in the comment in that PR.
